New relic testing support

ccobb-nr opened the pull request:
- Added X-Query-Key header support, which will allow running the tester directly against New Relics prometheus-api endpoint.
- Added ability to globally configure substitution in query strings, allowing the replacement of us to replace "_total" with "_total_cnt" in all configured queries at runtime.
- Added New Relic connection and query_tweak capabilities.

NOTE: The New Relic query_tweaks include ignore all "__name__" attributes in responses.  But, I hope to have that fixed on the New Relic side in the coming weeks.  That's why I've marked this as a draft.

ccobb-nr committed:
Add support for configuration of an X-Query-Key request header

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr committed:
Add support for global substitution in query strings

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr committed:
Add New Relic test_target_config and query_tweak comments

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

juliusv commented on cmd/promql-compliance-tester/main.go:
```diff
@@ -15,17 +16,28 @@ import (
+}
+
+func (rt RoundTripperWithHeader) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Add("X-Query-Key", rt.XQueryKey)
```
I had been looking at doing the same recently and saw that this would technically violate the [`RoundTripper` contract](https://golang.org/pkg/net/http/#RoundTripper):

```go
    // RoundTrip should not modify the request, except for
    // consuming and closing the Request's Body. RoundTrip may
    // read fields of the request in a separate goroutine. Callers
    // should not mutate or reuse the request until the Response's
    // Body has been closed.
```

I think this is ok for now, but it would be good to find a better solution long-term (unfortunately switching out the entire HTTP client in the Go Prometheus client would be more work), as it might lead to undefined behavior. At least we should add a `TODO` comment mentioning this.

ccobb-nr replied:
I'll add a comment.

juliusv commented on cmd/promql-compliance-tester/main.go:
```diff
@@ -15,17 +16,28 @@ import (
-	})
+func newPromAPI(targetConfig config.TargetConfig) (v1.API, error) {
+	apiConfig := api.Config{Address: targetConfig.QueryURL}
+	if len(targetConfig.XQueryKey) > 0 {
```
I know others will also want to set custom headers, so let's just allow configuring an arbitrary `map[string]string` vs. hardcoding a particular header name :)

ccobb-nr replied:
I started down that path, and went with the specific header because it was easier (and I don't know Go at all).  But I agree. I'll update this, but probably not for at least a few workdays.

juliusv commented on config/config.go:
```diff
@@ -18,8 +18,9 @@ type Config struct {
-	QueryURL string `yaml:"query_url"`
-	TSDBPath string `yaml:"tsdb_path"`
+	QueryURL  string `yaml:"query_url"`
+	XQueryKey string `yaml:"x_query_key"`
```
As mentioned above, I'd just call this `Headers` and allow passing arbitrary ones as a map.

ccobb-nr replied:
👍

juliusv commented on config/config.go:
```diff
@@ -30,6 +31,12 @@ type QueryTweak struct {
 	AlignTimestampsToStep  bool              `yaml:"align_timestamps_to_step" json:"alignTimestampsToStep,omitempty"`
 	DropResultLabels       []model.LabelName `yaml:"drop_result_labels" json:"dropResultLabels,omitempty"`
 	IgnoreFirstStep        bool              `yaml:"ignore_first_step" json:"ignoreFirstStep,omitempty"`
+	ApplyQuerySubstitution Substitution      `yaml:"apply_query_substitution" json:"applyQuerySubstitution,omitempty"`
```
How about just making this a map (old -> new)? That would then also allow doing multiple substitutions.

How about `ReplaceStringInQuery` or so for an even clearer name of what this option does?

ccobb-nr replied:
Yes, I will do that (probably next week). Thanks!

juliusv commented on testcases/expand.go:
```diff
@@ -124,6 +125,9 @@ func applyQueryTweaks(tc *comparer.TestCase, tweaks []*config.QueryTweak) *compa
 			resTC.Start = resTC.Start.Truncate(resTC.Resolution)
 			resTC.End = resTC.End.Truncate(resTC.Resolution)
 		}
+		if (config.Substitution{}) != t.ApplyQuerySubstitution {
```
(If I didn't suggest changing this to a map already, I would suggest changing optional query tweaks to a pointer so you could do the simpler `if t.ApplyQuerySubstitution != nil {` here.)

ccobb-nr replied:
You did.  But thanks for the Go help!  I never know when things will be nil vs. an empty map, etc.

ccobb-nr replied:
Done!

juliusv commented:
Really cool besides the comments, thanks! That's exactly the kind of contribution I was hoping for :)

mgmur committed:
Ability to add generic headers to http requests

Signed-off-by: Matthew Gmur <mgmur@newrelic.com>
Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr committed:
Remove New Relic drop_result_labels __name__ tweak

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr committed:
Change apply_query_substitution to query_string_replacements map.

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr committed:
Add TODO comment on RoundTrip header manipulation.

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr commented:
@juliusv I _think_ you and I have addressed your earlier concerns.  And, I've removed the `__name__` exclusion in the NR tweaks, as we are about to roll out a fix for that.  Promoting this PR to ready fro review. Let me know if I missed anything or you find anything else.

ccobb-nr committed:
Properly comment New Relic target config and query tweaks.

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

juliusv commented on promql-compliance-tester.yml:
```diff
@@ -30,6 +35,20 @@ query_tweaks:
+  #     - prometheus_server
+  # - note: 'New Relic omits the first resolution step in the output.'
+  #   ignore_first_step: true
+  # - note: 'New Relic stores metric counters as deltas, not cumulative values. New Relic will recognize these as guages by if "_cnt" is appended to the metric name.'
```
```suggestion
  # - note: 'New Relic stores metric counters as deltas, not cumulative values. New Relic will recognize these as gauges by if "_cnt" is appended to the metric name.'
```

juliusv commented on promql-compliance-tester.yml:
```diff
@@ -30,6 +35,20 @@ query_tweaks:
+  #   ignore_first_step: true
+  # - note: 'New Relic stores metric counters as deltas, not cumulative values. New Relic will recognize these as guages by if "_cnt" is appended to the metric name.'
+  #   query_string_replacements:
+  #     _total: _total_cnt
```
In the example test query configs for New Relic, I only replaced `_total` in some cases, namely where a counter metric was used in a way where we would look at its absolute value. I kept the normal `_total` suffix in queries that looked at the relative increase, e.g. by running `rate()` / `irate()` etc. on the counter metric. As far as I understand, this directive will just convert those latter cases to gauges as well, and then I'm not sure what NR's implementation of `rate()` etc. will yield when applied to gauges?

ccobb-nr replied:
I hadn't noticed that.  I will review the items that didn't append "_cnt" in your original customized NR config.

ccobb-nr replied:
I believe those relative increase methods only work with the guage type, which is what the _cnt extension causes the values to be interpreted as on the NR side.  

If I run the tester with just the relative increase queries, appending `_cnt` increases the success rate from 0% to 8.11%.  (Remaining failures, I believe, are unrelated to the `_cnt` extension.)

juliusv replied:
> I believe those relative increase methods only work with the guage type

This would strike me as odd, as those functions (`rate()` / `irate()` / `increase()` in Prometheus) are only meant to be used on counters, not on gauges (for those we have `deriv()`, `delta()`, etc.).

When I manually test `rate(demo_api_request_duration_seconds_count[1m])` vs. `rate(demo_api_request_duration_seconds_count_cnt[1m])`, the counter variant is also the one that looks closest to the native Prometheus:

**With counter metric in Prometheus (same for gauge, since Prometheus doesn't know the type, only looks at the value):**

![rate-on-prom-counter](https://user-images.githubusercontent.com/538008/96738966-938b6700-13bf-11eb-9f04-f725410a50cd.png)

**With counter-interpreted metric in New Relic:**

![rate_on_nr_counter](https://user-images.githubusercontent.com/538008/96738970-94bc9400-13bf-11eb-8a1e-3ad77f6976e7.png)

**With gauge-interpreted metric in New Relic:**

![rate_on_nr_gauge](https://user-images.githubusercontent.com/538008/96738972-94bc9400-13bf-11eb-9189-6b3f0590273b.png)

So we definitely want to interpret a counter-style metric as a New Relic counter when we plug it into `rate()` and friends.

juliusv replied:
> If I run the tester with just the relative increase queries, appending _cnt increases the success rate from 0% to 8.11%.

Following up on this as well, I wonder whether we are talking about the same test cases. The relevant test case for `rate`/`irate`/`increase` (all the counter-related increase functions) would be this one:

```yaml
  - query: '{{.extrapolatedRateFunc}}(demo_cpu_usage_seconds_total[{{.range}}])'
    variant_args: ['extrapolatedRateFunc', 'range']
```

But I get a 0% passing rate for that, no matter if I turn on the query replacement query tweak or manually append `_cnt` or just keep it as it is. (Still, as seen above, keeping it a counter-interpreted metric for those functions leads to a visually closer result to Prometheus)

juliusv replied:
(I am running this against my personal NR account by the way, so not sure if you're already testing with some account that has more features or fixes enabled).

mgmur replied:
@juliusv @ccobb-nr I've added an "exemptions" to the yaml, where functions can be added to exempt them from the metric name substitution. Addressed in [3f2dff853d4ce84ad52255825b64d0b18118cb18](https://github.com/promlabs/promql-compliance-tester/pull/2/commits/3f2dff853d4ce84ad52255825b64d0b18118cb18).

juliusv commented on promql-compliance-tester.yml:
```diff
@@ -30,6 +35,20 @@ query_tweaks:
+  # - note: 'New Relic is sometimes off by 1ms when parsing floating point start/end timestamps.'
+  #   truncate_timestamps_to_ms: 1000
+  # - note: 'New Relic adds a "prometheus_server" label to distinguish Prometheus servers, leading to extra labels in query results. These need to be stripped before comparisons.'
+  # - note: 'New Relic does not consistently include the "__name__" attribute in its responses.'
```
This line is still left over from the `__name__` query tweak that was removed.

juliusv reviewed (commented):
Looks great besides some nits! I'm not sure what to do about the counter replacement comment yet, but wanted to point it out.

ccobb-nr committed:
Correct spelling of gauges in promql-compliance-tester.yml

Co-authored-by: Julius Volz <julius.volz@gmail.com>
Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr committed:
Remove New Relic __name__ note.

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

mgmur committed:
adds a way to define exemptions for query substitutions for specific functions

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

mgmur committed:
query substitution bug fix

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

juliusv commented on testcases/expand.go:
```diff
@@ -124,6 +125,14 @@ func applyQueryTweaks(tc *comparer.TestCase, tweaks []*config.QueryTweak) *compa
 		}
+		if len(t.QueryStringReplacements.Remove) > 0 {
+			for _, e := range t.QueryStringReplacements.Exemptions {
+				if strings.HasPrefix(resTC.Query, e) {
```
Hmm, I think this query tweak implementation is ok as a local stop-gap for your testing, but it's a bit brittle and special-cased for putting it like this in the upstream tester tool. Especially as it only checks the prefix of queries, whereas functions may appear anywhere within a test query string, and the set of test queries is subject to change and expansion over time (so e.g. `rate()` will appear deeper within queries as well, or similar-looking strings will not actually be function calls).

I see two options as nicer alternatives, each of which I'm happy to take on / help with:

* A) Change the set of test queries to not look at absolute values of counters in the first place, use gauges instead. With the exception of a single test case that explicitly tests whether the absolute value of a counter can be retrieved. The main reason why I'm currently using counter metrics in many of the non-`rate()`-y test cases is that the example counter metrics that are exposed by the demo application have a good set of labels to experiment on. If we changed the test dataset to have equally useful gauges for testing, we could just switch those cases over (no replacement magic needed anymore then), and I would even see that as an overall improvement to the test cases to not query so much for absolute values of counters (although for PromQL-compliant implementations, that shouldn't matter).
* B) Add a more robust and general query replacement mechanism that does a full parse of the expression (https://github.com/prometheus/prometheus/blob/63be30dceed942ac53c0cc6d8637e6563598cd87/promql/parser/parse.go#L105), followed by an iteration through the query nodes to apply tweaks in a robust fashion (https://github.com/prometheus/prometheus/blob/63be30dceed942ac53c0cc6d8637e6563598cd87/promql/parser/ast.go#L259-L306), then re-serialize the expression. This is easier than it sounds :) Here is an example of someone parsing an expression and then doing that kind of iteration to enforce the presence of a certain set of labels on selectors: https://github.com/prometheus-community/prom-label-proxy/blob/0211f3ff373505b4bd3368f8c3f9a1962580578b/injectproxy/routes.go#L120-L142. **But the challenge I see this this option is to provide a general-enough way of specifying what to mutate where in the tree, because other people will have similar, but different needs, such as enforcing a label selector, or such.**

I think both alternatives could be actual enrichments to the tester tool, but only one is needed for NR's purposes. 

The first one would be easiest, let me know if you're happy with that and I can get right on that (I might do so in any case).

ccobb-nr replied:
@juliusv I think your solution A sounds like a good plan.  Can you go ahead and implement that?

juliusv commented on promql-compliance-tester.yml:
```diff
@@ -30,6 +35,24 @@ query_tweaks:
+  #     remove: _total
+  #     replacement: _total_cnt
+  #     exemptions:
+  #       - delta
```
I think this should be `irate` and not `delta`, as `delta` is meant to be used on gauges, while `irate` is another counter function that is missing in this list.

Btw. `resets` is another counter-only function, but I guess that one will not be able to work anyway as long as NR doesn't store absolute counter values that would allow computing where a counter reset happened.

ccobb-nr committed:
Rebase with updated tester config, using demo_memory_usage_bytes metric.

Signed-off-by: Craig Cobb <ccobb@newrelic.com>
This commit removes the no longer needed string replacement tweak.
Signed-off-by: Craig Cobb <ccobb@newrelic.com>

ccobb-nr committed:
Test decimal without leading zero. Test standard scientific notation.

Signed-off-by: Craig Cobb <ccobb@newrelic.com>

juliusv commented:
:+1: All looking good now, thanks! Could you just ensure that all commits have the DCO signoff (via interactive rebase or such)?

DCO check says:

```
Commit sha: 06c8923, Author: ccobb-nr, Committer: Craig Cobb; The sign-off is missing.
Commit sha: 0822156, Author: Matthew Gmur, Committer: Craig Cobb; The sign-off is missing.
Commit sha: 0088bb5, Author: Matthew Gmur, Committer: Craig Cobb; The sign-off is missing.
Commit sha: b6b7694, Author: Craig Cobb, Committer: Craig Cobb; The sign-off is missing.
```

ccobb-nr commented:
@juliusv Ping! I did the rebase, DCO issue cleared.  :)

juliusv commented:
:+1: Awesome, thanks :)
